src/Parlot/Fluent/Then.cs

```diff
@@ -6,49 +6,156 @@
+
             var parsed = new ParseResult<T>();
 
             if (_parser.Parse(context, ref parsed))
+            {
+                if (_transform1 != null)
+                {
+                    result.Set(parsed.Start, parsed.End, _transform1(parsed.Value));
+                }
+                else if (_transform2 != null)
+                {
+                    result.Set(parsed.Start, parsed.End, _transform2(context, parsed.Value));
+                }
+
+                return true;
+            }
+
+            return false;
+        }
+    
+        public CompilationResult Compile(CompilationContext<TParseContext> context)
+        {
+            var result = new CompilationResult();
+
+            var success = context.DeclareSuccessVariable(result, false);
+            var value = context.DeclareValueVariable(result, Expression.Default(typeof(U)));
+
+            // parse1 instructions
+            // 
+            // if (parser1.Success)
+            // {
+            //    success = true;
+            //    value = action(parse1.Value);
+            // }
+
+            var parserCompileResult = _parser.Build(context, requireResult: true);
+
+            Expression transformation;
+
+            if (_transform1 != null)
+            {
+                transformation = Expression.Invoke(Expression.Constant(_transform1), new [] { parserCompileResult.Value });
+            }
+            else if (_transform2 != null)
+            {
+                transformation = Expression.Invoke(Expression.Constant(_transform2), context.ParseContext, parserCompileResult.Value);
+            }
+            else
+            {
+                transformation = Expression.Default(typeof(U));
+            }
+
+            var block = Expression.Block(
+                    parserCompileResult.Variables,
+                    parserCompileResult.Body
+                    .Append(
+                        Expression.IfThen(
+                            parserCompileResult.Success,
+                            Expression.Block(
+                                Expression.Assign(success, Expression.Constant(true, typeof(bool))),
+                                context.DiscardResult
+                                ? Expression.Empty()
+                                : Expression.Assign(value, transformation)
+                                )
+                            )
+                        )
+                    );
+
+            result.Body.Add(block);
+
+            return result;
+        }
+    }
+
+     /// <summary>
+    /// Returns a new <see cref="Parser{U,TParseContext}" /> converting the input value of 
+    /// type T to the output value of type U using a custom function.
+    /// </summary>
+    /// <typeparam name="T">The input parser type.</typeparam>
+    /// <typeparam name="TParseContext">The parse context type.</typeparam>
+    public sealed class Then<T, TParseContext> : Parser<T, TParseContext>, ICompilable<TParseContext>
```

Comment:
That's a new one right?
I assume to just be able to pass an action that can change the context only?

Reply:
indeed